Blog post error styles (#44)
pypy/pypy.org: posts/2021/12/error-message-style-guides.txt

.. title: Error Message Style Guides of Various Languages
.. slug: error-message-style-guides
.. date: 2021-12-10 16:00:00 UTC
.. tags:
.. category: 
.. link: 
.. description: 
.. type: rest
.. author: Carl Friedrich Bolz-Tereick

================================================
Error Message Style Guides of Various Languages
================================================

PyPy has been trying to produce good `SyntaxErrors`_ and `other`_ `errors`_ for

> a long time, and thankfully CPython has made an enormous push to `improve its
> SyntaxErrors in the last releases`_. These improvements are great, but the process

Reviewer: ```suggestion
a long time. CPython has also made an enormous push to `improve its
SyntaxErrors in the last few releases`_. These improvements are great, but the process
```

Wonderfully, people answered me with lots of helpful links (`full list`_ at the
end of the post), thank you everybody! All those sources are very interesting
and contain many great points, I recommend reading them directly! In this
post, I'll try to summarize some common themes or topics that I thought were
particularly interesting.

Language Use
------------

Almost all guides stress the need for plain and simple English, as well as
conciseness and clarity [Flix, Racket, Rust, Flow]. Flow suggests to put coding
effort into making the grammar correct, for example in the case of plurals or
to distinguish between "a" and "an".

The suggested tone should be friendly and neutral, the messages should not
blame the Programmer [Flow]. Rust and Flix suggest to not use the term
'illegal' and use something like 'invalid' instead.

Flow suggests to avoid "compiler speak". For example terms like 'token' and
'identifier' should be avoided and terms that are more familiar to programmers
be used (eg "name" is better). The Racket guide goes further and has a list of
allowed technical terms and some prohibited terms.

Structure
---------

> Several guides (such as Flix and Flow) point out a 80/20 rule: 80% of the times an error message is
> read, the developer knows that message well and knows exactly what to do. For

Reviewer: ```suggestion
Several guides (such as Flix and Flow) point out a 80/20 rule: 80% of the times
the developer knows that message well and knows exactly what to do. For
```

this use case it's important that the message is short. On the other hand, 20%
of the times this same message will have to be understood by a developer who
has never seen it before and is confused, and so the message needs to contain
enough information

> to allow them to find out what is going on.

Reviewer: A few examples might drive home the point. Some kind of actionable continuation of this thought might be nice. Can we distinguish between the two types of errors with a comment in the code where the error is produced (`#common`, `#rare`)? Do we need an "beginner" or "advanced" environment variable to produce different messages? Or is this thought needed for the next paragraph, in which case it would be nice to point out how the 80/20 principle guides this structure, to connect the ideas.

Author: Yeah, the thing that I have a hard time getting across (and I think the sources too, but well) is that the point is that it's *the same error* that can be either part of the 80% or the 20%. Eg if I (or another experienced python dev) see this:

```
>>> a.f(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: f() takes 1 positional argument but 2 were given
```
I know that I forgot the `self` parameter in the method definition. But a beginner (the 20% case) will be extremely confused by that same message. So I think what that rule of thumb wants to say is that every message exists in this tension between "keep the error short, the devs don't read it properly" and "give a lot of information for people that see it the very first time".

The Racket guide proposes to use the following general structure for errors:
'State the constraint that was violated ("expected a"), followed by what was
found instead.'

The Rust guides says to avoid "Did you mean?" and questions in general, and
wants the compiler to instead be explicit why something was suggested. The
example the Rust guide gives is: 'Compare "did you mean: Foo" vs. "there is a
struct with a similar name: Foo".' Racket goes further and forbids
suggestions altogether because "Students will follow well‐meaning‐but‐wrong
advice uncritically, if only because they have no reason to doubt the

> authoritative voice of the tool."

Reviewer: Hmm. What do you think about this? Python has taken a different direction and people seem to like it. Should suggestions be left to people using IDEs where the UX is different?

Author: I don't know, I indeed find the second part too strict somehow, but it's still interesting (and probably the only of the guides that is backed by any kind of real user research). I like the idea of rephrasing "Did you mean" to "Here is something that could fit *because* ...". That gives more information why the suggestion was given.

Reviewer: As for the text, the advantage of "Did you mean" is that it is lightweight at 12 letters.

Maybe some more tools should be tried: type annotations or some other way to figure out better candidates.

Formatting and Source Positions
-------------------------------

> The Rust guide suggests to put all identifiers into backticks, Flow formats the error messages using full Markdown.

Reviewer: Different thoughts, different sentences. Maybe different paragraphs?
```suggestion
The Rust guide suggests to put all identifiers into backticks.

Flow formats the error messages using full Markdown.
```

Author: There is a connection and I think I should make that explicit: backticks are markdown for code snippets.

conciseness and clarity [Flix, Racket, Rust, Flow]. Flow suggests to put coding
effort into making the grammar correct, for example in the case of plurals or
to distinguish between "a" and "an".

The suggested tone should be friendly and neutral, the messages should not
blame the Programmer [Flow]. Rust and Flix suggest to not use the term
'illegal' and use something like 'invalid' instead.

Flow suggests to avoid "compiler speak". For example terms like 'token' and
'identifier' should be avoided and terms that are more familiar to programmers
be used (eg "name" is better). The Racket guide goes further and has a list of
allowed technical terms and some prohibited terms.

Structure
---------

Several guides (such as Flix and Flow) point out a 80/20 rule: 80% of the times an error message is
read, the developer knows that message well and knows exactly what to do. For
this use case it's important that the message is short. On the other hand, 20%
of the times this same message will have to be understood by a developer who
has never seen it before and is confused, and so the message needs to contain
enough information

> to allow them to find out what is going on.

Reviewer: Maybe something like this to close out the paragraph
```suggestion
to allow them to find out what is going on. So the error message needs to strike
a balance between brevity and clarity.
```
